Fixed #31181 -- Added links to related models for admin's readonly fields.

julienr opened the pull request:
In the django admin, readonly relationship fields (e.g. ForeignKey, ManyToMany) were rendered as plaintext of the target object.
This changes this to render them as links instead, making it possible to navigate through the relationship.

## Before
<img width="200" alt="Screenshot 2020-01-21 at 21 59 59" src="https://user-images.githubusercontent.com/506602/72842969-e356c980-3c99-11ea-8c7a-9b70e9689efb.png">

## After
<img width="200" alt="Screenshot 2020-01-21 at 22 00 12" src="https://user-images.githubusercontent.com/506602/72842981-e81b7d80-3c99-11ea-82b0-925cf8e0cebd.png">



Based on [feedback](https://groups.google.com/forum/#!topic/django-developers/XFoaohDpqZE) / [discussion](https://github.com/julienr/django/commit/01bba0b714f831a669b271c1192e93fc3ef011b8#commitcomment-36862302) with @aaugustin , things to decide:

- [x] Should this stop rendering links for ManyToMany if there are more than say 20 remote objects ? The current plaintext behavior already renders all objects ? Is the concern here the number of calls to `reverse` ? Should I make 20 a config option or just hardcode it ?

- [x] The link should only be rendered if the user has permission to view the target model. This would be nice, but the current `master` behavior with raw_id fields is to render the link even if the user cannot view the target model. So I feel changing this would be another issue. I looked a bit into it and it seems fixing this would require passing the `request` all the way down to be able to check for permissions. Basically from `ModelAdmin._changeform_view` =>  `AdminForm` => `Fieldset` => `Fieldline` => `AdminReadonlyField`. Should I open a separate ticket for this ?

- [x] In the [contribution guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist), it says new features should have an entry in the release notes. Is this considered a new feature ? I feel this is pretty minor, but not sure what the policy is

julienr commented:
> Should this stop rendering links for ManyToMany if there are more than say 20 remote objects ?

The current master behavior is to render all the objects as plaintext, separated by commas. Did some performance testing, measuring the time to load the page (from the browser dev console).

With 1000 related objects, average of 5 runs:
- 80-90ms for master
- 270-300ms for this PR

Digging a bit, this is indeed the call to `reverse` that's causing the slowdown.

I guess one option would be to stop rendering links after the 20th object, so it would look like this:

<img width="1153" alt="Screenshot 2020-01-20 at 21 40 07" src="https://user-images.githubusercontent.com/506602/72756424-783fc100-3bcd-11ea-99ff-70c08ece4d07.png">


I'm just wondering if I should hardcode the 20 of make it configurable ?

inoks commented on django/contrib/admin/helpers.py:
```diff
@@ -213,7 +231,13 @@ def contents(self):
                 if isinstance(f.remote_field, ManyToManyRel) and value is not None:
-                    result_repr = ", ".join(map(str, value.all()))
+                    result_repr = []
+                    for i, remote_obj in enumerate(value.all()):
```
What do we need to have `enumerate()` here?

julienr replied:
Good catch ! That was a leftover - removed it

felixxm commented:
IMO we shouldn't render links for m2m fields, this will regress performance. Also, rendering the first 20 links is misleading and don't have much value. I would like to limit this ticket only to foreign keys.

carltongibson commented:
Yes, the links aren't worth the performance regression. I think skip them on this pass.

julienr commented:
Thanks for the feedback! I've amended the PR to only include ForeignKey/OneToOne relationships. 

Would expanding this to ManyToMany fields be something you would consider if I can come up with a solution that doesn't impact performance as much ? Or maybe just have a way for the user to enable/disable this through the ModelAdmin (and we'd disable by default) ? Maybe something to discuss in another ticket ?

carltongibson commented:
Hi @julienr. 

> Would expanding this to ManyToMany fields... 

We have enough admin options. No more of those. :)

I'd like to suggest that folks can just use a custom widget here. But this touches the read-only rendering, so without thinking it through I'm not immediately sure what that entails. Folks should be able to override something to get the behaviour they want here. If you fancy looking at what that would entail, we can discuss whether the right extension points are in place. 

But a separate ticket, yes.

julienr commented:
> We have enough admin options. No more of those. :)

Ahahah, I understand the pain ;)

> I'd like to suggest that folks can just use a custom widget here. But this touches the read-only rendering, so without thinking it through I'm not immediately sure what that entails. Folks should be able to override something to get the behaviour they want here. If you fancy looking at what that would entail, we can discuss whether the right extension points are in place.

Right, I'll think/research about this and open a ticket if I find a feasible solution.

julienr commented:
Does anyone have time to review this ? Anything else I can do to improve it ? Thanks :)

carltongibson commented:
We’ll get to it. Thanks for you effort and patience.

felixxm commented on django/contrib/admin/helpers.py:
```diff
@@ -189,6 +192,21 @@ def label_tag(self):
+                ),
+                args=(remote_obj.id,)
+            )
+            link = format_html('<a href="{}">{}</a>', url, remote_obj)
+            return link
```
```suggestion
            return format_html('<a href="{}">{}</a>', url, remote_obj)
```

julienr replied:
Fixed in my last commit

felixxm commented on django/contrib/admin/helpers.py:
```diff
@@ -189,6 +192,21 @@ def label_tag(self):
+            link = format_html('<a href="{}">{}</a>', url, remote_obj)
+            return link
+        except NoReverseMatch:
+            # This handles the case where the remote model has no admin
```
We can chop this comment.

julienr replied:
Fixed in my last commit

felixxm commented on django/contrib/admin/helpers.py:
```diff
@@ -189,6 +192,21 @@ def label_tag(self):
+            url = reverse(
+                'admin:%s_%s_change' % (
+                    remote_field.model._meta.app_label,
+                    remote_field.model._meta.object_name.lower()
```
```suggestion
                    remote_field.model._meta.model_name,
```

julienr replied:
Fixed in my last commit

felixxm commented on django/contrib/admin/helpers.py:
```diff
@@ -214,6 +232,9 @@ def contents(self):
             else:
                 if isinstance(f.remote_field, ManyToManyRel) and value is not None:
                     result_repr = ", ".join(map(str, value.all()))
+                elif (isinstance(f.remote_field, ForeignObjectRel) or isinstance(f.remote_field, OneToOneField)) \
+                        and value is not None:
```
```suggestion
                elif (
                    isinstance(f.remote_field, (ForeignObjectRel, OneToOneField)) and
                    value is not None
                ):
```

felixxm replied:
I think this should take into account the `parent_link` attribute for `OneToOneField`s :thinking:

julienr replied:
Thank you very much for the review. I'll fix the other issues, but I'm not sure what is the expected behavior for `parent_link`:

It seems if I have the following
```
class ModelParent(models.Model):
    name = models.CharField(max_length=60)

class ModelChild(ModelParent):
    parent = models.OneToOneField(ModelParent, models.CASCADE, primary_key=True, parent_link=True)
```

The the behavior of the admin on master is to not display `parent` and instead display the parent's `name` directly as a field in the admin form for `ModelChild`.

So I feel there isn't anything special to do for `OneToOneField` with `parent_link=True` since those don't seem to be displayed anyway.

I might very well be missing something and I have to say I'm not super familiar with `parent_link=True`, so maybe I misunderstand the expected behavior. What did you have in mind for the behavior with `parent_link` ?

felixxm replied:
True, I was thinking that maybe it's worth to cover them in tests, but you're right be don't need to test them separately.

julienr replied:
Ok thanks !

felixxm commented on tests/admin_views/admin.py:
```diff
@@ -539,6 +539,10 @@ class PizzaAdmin(admin.ModelAdmin):
 
 
+class ReadonlyModelAdmin(admin.ModelAdmin):
+    readonly_fields = ('model_with_admin', 'model_without_admin', )
```
```suggestion
    readonly_fields = ('model_with_admin', 'model_without_admin')
```

julienr replied:
Fixed in my last commit

felixxm commented on tests/admin_views/models.py:
```diff
@@ -999,3 +999,23 @@ class UserProxy(User):
+class ReadonlyModel(models.Model):
+    name = models.TextField()
+    model_with_admin = models.ForeignKey(ModelWithAdmin, models.CASCADE)
+    model_without_admin = models.ForeignKey(ModelWithoutAdmin, models.CASCADE)
```
You can reuse existing models, e.g. `ChapterXtra1` and `Vodcast`.

julienr replied:
Fixed in my last commit

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -4899,6 +4900,25 @@ def test_change_form_renders_correct_null_choice_value(self):
 
+    def test_readonly_foreignkey(self):
+        """
+        Test that ForeignKey renders as links for readonly fields if the target model
```
You should omit prefixes like "Test that" since all tests ensure things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

julienr replied:
Fixed in my last commit

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -4899,6 +4900,25 @@ def test_change_form_renders_correct_null_choice_value(self):
+        self.assertContains(response, '<div class="readonly"><a href="%s">model with admin</a></div>' % link_to_book,
+                            html=True)
+        # model without admin should be rendered as plaintext
+        self.assertContains(response, '<div class="readonly">model without admin</div>', html=True)
```
Please add tests for `OneToOneField` with `parent_link` and admin, you can use the `Vodcast` model.

julienr replied:
As per our discussion above, this doesn't seem needed.

felixxm reviewed (commented):
@julienr Thanks for this patch :+1: 

Please add also small release notes.

julienr commented:
I think I've addressed the comments on the PR. Tomorrow, I'll give this another test round on our internal codebase at work that has a pretty heavy & customized admin to make sure this doesn't break anything.

julienr commented:
> I think I've addressed the comments on the PR. Tomorrow, I'll give this another round of manual tests on our internal codebase at work that has a pretty heavy & customized admin to make sure the latest changes aren't breaking things.

Seems to work fine.

julienr commented:
@felixxm I think I've addressed your comments in a new commit (that we can then squash). If you can have a look again and let me know what you think, that would be great.

For release notes, I've added this

> Read-only relationship fields are now rendered as navigable links to the related model.

I'm not sure if this needs more than this or not (seems like a small feature)

pope1ni commented on django/contrib/admin/helpers.py:
```diff
@@ -187,6 +190,19 @@ def label_tag(self):
+    def _get_link_to_related_model(self, remote_field, remote_obj):
+        try:
+            url = reverse(
+                'admin:%s_%s_change' % (
+                    remote_field.model._meta.app_label,
+                    remote_field.model._meta.model_name
+                ),
+                args=(remote_obj.id,)
```
```suggestion
                'admin:%s_%s_change' % (
                    remote_field.model._meta.app_label,
                    remote_field.model._meta.model_name,
                ),
                args=[remote_obj.pk],
```

julienr replied:
Fixed in my last commit

pope1ni commented on docs/releases/3.2.txt:
```diff
@@ -71,6 +71,8 @@ Minor features
 
 * :attr:`.ModelAdmin.search_fields` now allows searching against quoted phrases
   with spaces.
+* Read-only relationship fields are now rendered as navigable links to the
```
Please add a blank line above.

julienr replied:
Fixed in my last commit

pope1ni commented on tests/admin_views/tests.py:
```diff
@@ -5042,6 +5042,27 @@ def test_change_form_renders_correct_null_choice_value(self):
+        # User has an admin, Chapter doesn't
+        user = User.objects.create_user(username='testuser', password='secret')
+        book = Book.objects.create(name='testbook')
+        chapter = Chapter.objects.create(title="testchapter", book=book)
```
```suggestion
        chapter = Chapter.objects.create(title='testchapter', book=book)
```

julienr replied:
Fixed in my last commit

felixxm reviewed (commented):
@julienr Thanks for updates :+1: Welcome aboard :boat: 

I pushed edits and fixed two issues with custom primary keys.

julienr commented on django/contrib/admin/helpers.py:
```diff
@@ -187,6 +190,19 @@ def label_tag(self):
+                    remote_field.model._meta.app_label,
+                    remote_field.model._meta.model_name
+                ),
+                args=[remote_obj.id],
```
Good catch !

julienr commented:
Thanks @felixxm for your help getting this through !

Skrattoune commented:
Hi, I was very excited by this new feature.
I got it working quite fine on ChangeForm pages and in Inlines,
but it does not seem to work for me on ChangeList pages.
That's for me key for easy navigation.

Is it normal?
If so, why wasn't it implemented?

Thanks in advance

julienr commented:
@Skrattoune good question !

> I got it working quite fine on ChangeForm pages and in Inlines, but it does not seem to work for me on ChangeList pages.

Indeed, this wasn't implemented, but I don't think there was a particularly strong reason not to do it. I would be happy to take a stab at it if you [report it as a bug/feature first](https://docs.djangoproject.com/en/dev/internals/contributing/bugs-and-features/) and nobody has strong objections against this.

lucasmendes2105 commented:
@julienr The idea is good but it should be optional.
There are several cases where it is not interesting to insert the link.

sguldemond commented:
@julienr Is there a way to diable this behavior for specific ForeignKey fields?

julienr commented:
> @julienr Is there a way to disable this behavior for specific ForeignKey fields?

I think if the target model isn't registered in the admin, that would disable this. But there is no other/explicit option in the current state.
